ludwig/schema/config_object.py

```diff
@@ -0,0 +1,469 @@
+
+
+class BaseFeatureContainer:
+    """Base Feature container for input and output features."""
+
+    def to_dict(self):
+        """Method for getting a dictionary representation of the input features.
+
+        Returns:
+            Dictionary of input features specified.
+        """
+        return convert_submodules(self.__dict__)
+
+    def to_list(self):
+        """Method for getting a list representation of the input features.
+
+        Returns:
+            List of input features specified.
+        """
+        return list(convert_submodules(self.__dict__).values())
+
+    def filter_features(self):
+        return {
+            key: {k: v for k, v in value.items() if k in {NAME, TYPE, ACTIVE}} for key, value in self.to_dict().items()
+        }
+
+    def __repr__(self):
+        filtered_repr = self.filter_features()
+        return yaml.dump(filtered_repr, sort_keys=True)
+
+
+class InputFeaturesContainer(BaseFeatureContainer):
+    """InputFeatures is a container for all input features."""
+
+    pass
+
+
+class OutputFeaturesContainer(BaseFeatureContainer):
+    """OutputFeatures is a container for all output features."""
+
+    pass
+
+
+@dataclass(repr=False)
+class Config(BaseMarshmallowConfig):
+    """This class is the implementation of the config object that replaces the need for a config dictionary
+    throughout the project."""
+
+    def __init__(self, config_dict: dict):
+
+        # ===== Backwards Compatibility =====
+        upgraded_config = upgrade_to_latest_version(config_dict)
+
+        # ===== Initialize Top Level Config Sections =====
+        self.model_type: str = MODEL_ECD
+        self.input_features: InputFeaturesContainer = InputFeaturesContainer()
+        self.output_features: OutputFeaturesContainer = OutputFeaturesContainer()
+        self.combiner: BaseCombinerConfig = ConcatCombinerConfig()
+        self.trainer: BaseTrainerConfig = ECDTrainerConfig()
+        self.preprocessing: PreprocessingConfig = PreprocessingConfig()
+        self.defaults: DefaultsConfig = copy.deepcopy(DefaultsConfig())
+
+        # ===== Set User Defined Global Defaults =====
+        if DEFAULTS in upgraded_config:
+            self._set_attributes(self.defaults, upgraded_config[DEFAULTS])
+
+        # ===== Features =====
+        self._set_feature_column(upgraded_config)
+        self._set_proc_column(upgraded_config)
+        self._parse_features(upgraded_config[INPUT_FEATURES], INPUT_FEATURES)
+        self._parse_features(upgraded_config[OUTPUT_FEATURES], OUTPUT_FEATURES)
+
+        # ===== Model Type =====
+        if MODEL_TYPE in upgraded_config:
+            if upgraded_config[MODEL_TYPE] == MODEL_GBM:
+                self.model_type = MODEL_GBM
+                self.trainer = GBMTrainerConfig()
+                if TYPE in upgraded_config.get(TRAINER, {}):
+                    assert upgraded_config[TRAINER][TYPE] == "lightgbm_trainer"
```

Comment:
`raise ValidationError()` (assert will raise AssertionError).  Consistent exception type will make it easier for callers to catch validation errors.

Reply:
Good catch. Changed!